### ISSUE TITLE
说说什么是Transformer中的Multi head attention？

### ISSUE THREAD

**Issue by Sogrey** [Sogrey/AI-QA#386]
## 📌 什么是Transformer中的Multi-Head Attention？

Multi-Head Attention（多头注意力机制）是Transformer模型中的一项核心技术，它通过多个并行的注意力头（Head）来捕捉输入序列中不同位置之间的依赖关系。每个注意力头独立地计算输入序列中不同位置的注意力权重，最终将这些头的结果进行拼接和线性变换，以得到最终的输出。

### 🧠 **1. 原理与结构**

- **单头注意力（Single-Head Attention）**：
 - 通过查询（Query）、键（Key）和值（Value）三个向量进行计算，公式如下：
 $$ \text{Attention}(Q, K, V) = \text{softmax}(\frac{QK^T}{\sqrt{d_k}})V $$
 - 其中 $ d_k $ 是键向量的维度。

- **多头注意力**：
 - 将输入向量通过线性变换生成多个不同的 $ Q $、$ K $、$ V $，每个头独立计算注意力权重。
 - 最终将所有头的结果进行拼接（Concat）并应用一个线性变换，以得到最终输出。

### 📈 **2. 优势**

- **捕捉多尺度依赖关系**：每个头可以关注输入序列的不同部分，从而增强模型对复杂模式的捕捉能力。
- **提升模型表达能力**：通过并行计算多个注意力头，模型可以更全面地理解上下文信息。
- **并行计算**：多头注意力机制支持高效的并行计算，有助于提高模型的训练和推理速度。

### 🧩 **3. 应用场景**

- **自然语言处理（NLP）**：如机器翻译、文本摘要、问答系统等。
- **计算机视觉（CV）**：如图像识别、目标检测等。
- **语音识别**：通过捕捉语音信号中的时序信息。

### 🧪 **4. 示例代码（PyTorch）**

```python
import torch
import torch.nn as nn

class MultiHeadAttention(nn.Module):
 def __init__(self, d_model, num_heads):
 super(MultiHeadAttention, self).__init__()
 self.num_heads = num_heads
 self.d_model = d_model
 self.d_k = d_model // num_heads
 self.W_q = nn.Linear(d_model, d_model)
 self.W_k = nn.Linear(d_model, d_model)
 self.W_v = nn.Linear(d_model, d_model)
 self.W_o = nn.Linear(d_model, d_model)

 def forward(self, Q, K, V):
 batch_size = Q.size(0)
 Q = self.W_q(Q).view(batch_size, -1, self.num_heads, self.d_k).transpose(1, 2)
 K = self.W_k(K).view(batch_size, -1, self.num_heads, self.d_k).transpose(1, 2)
 V = self.W_v(V).view(batch_size, -1, self.num_heads, self.d_k).transpose(1, 2)
 
 scores = torch.matmul(Q, K.transpose(-2, -1)) / \sqrt(self.d_k)
 scores = torch.softmax(scores, dim=-1)
 output = torch.matmul(scores, V)
 output = output.transpose(1, 2).contiguous().view(batch_size, -1, self.d_model)
 output = self.W_o(output)
 return output

# 示例使用
Q = torch.randn(2, 5, 10)
K = torch.randn(2, 5, 10)
V = torch.randn(2, 5, 10)
attention = MultiHeadAttention(10, 2)
result = attention(Q, K, V)
print(result.shape)
```

### ✅ **总结**

Multi-Head Attention 通过多个并行的注意力头，增强了模型对输入序列中不同位置之间依赖关系的捕捉能力，是Transformer模型的核心组成部分。它在自然语言处理、计算机视觉等多个领域中广泛应用，并通过高效的并行计算提升了模型的性能和表达能力。

======

来自n8n
